File: tui/src/scala/tui/widgets/canvas/CanvasWidget.scala
package tui
package widgets
package canvas

import tui.Style
import tui.internal.ranges

/** The Canvas widget may be used to draw more detailed figures using braille patterns (each cell can have a braille character in 8 different positions).
  *
  * @param block
  * @param xBounds
  * @param yBounds
  * @param backgroundColor
  * @param marker
  *   Change the type of points used to draw the shapes. By default the braille patterns are used as they provide a more fine grained result but you might want
  *   to use the simple dot or block instead if the targeted terminal does not support those symbols.
  * @param painter
  */
case class CanvasWidget(
    xBounds: Point = Point.Zero,
    yBounds: Point = Point.Zero,
    backgroundColor: Color = Color.Reset,
    marker: symbols.Marker = symbols.Marker.Braille
)(painter: Context => Unit)
    extends Widget {
  override def render(area: Rect, buf: Buffer): Unit = {
    buf.setStyle(area, Style.DEFAULT.bg(this.backgroundColor))

    // Create a blank context that match the size of the canvas
    val ctx = Context(
      area.width,
      area.height,
      this.xBounds,
      this.yBounds,
      this.marker
    )
    // Paint to this context
    painter(ctx)
    ctx.finish()

    // Retreive painted points for each layer
    ctx.layers.foreach { layer =>
      ranges.range(0, math.min(layer.string.length, layer.colors.length)) { i =>
        val ch = layer.string.charAt(i)
        val color = layer.colors(i)
        if (ch != ' ' && ch != '\u2800') {
          val (x, y) = (i % area.width, i / area.width)
          buf
            .get(x + area.left, y + area.top)
            .setChar(ch)
            .setFg(color)
          ()
        }
      }
    }

    // Finally draw the labels
    val left = this.xBounds.x
    val right = this.xBounds.y
    val top = this.yBounds.y
    val bottom = this.yBounds.x
    val width = math.abs(this.xBounds.y - this.xBounds.x)
    val height = math.abs(this.yBounds.y - this.yBounds.x)
    val resolution = {
      val width = (area.width - 1).toDouble
      val height = (area.height - 1).toDouble
      (width, height)
    }
    ranges.range(0, ctx.labels.length) { i =>
      val l = ctx.labels(i)
      if (l.x >= left && l.x <= right && l.y <= top && l.y >= bottom) {
        val label = l
        val x = ((label.x - left) * resolution._1 / width).toInt + area.left
        val y = ((top - label.y) * resolution._2 / height).toInt + area.top
        buf.setSpans(x, y, label.spans, area.right - x)
        ()
      }
    }
  }
}

File: tui/src/java/tui/widgets/canvas/CanvasWidget.java
package tui.widgets.canvas;

import java.util.Optional;
import java.util.function.Consumer;
import tui.Buffer;
import tui.Color;
import tui.Point;
import tui.Rect;
import tui.Style;
import tui.Symbols;
import tui.Widget;
import tui.internal.Ranges;
import tui.widgets.BlockWidget;

/// The Canvas widget may be used to draw more detailed figures using braille patterns (each cell can have a braille character in 8 different positions).
public final class CanvasWidget implements Widget {
  public final Optional<BlockWidget> block;
  public final Point xBounds;
  public final Point yBounds;
  public final Color backgroundColor;
  public final Symbols.Marker marker;
  public final Consumer<Context> painter;

  public CanvasWidget(
      Optional<BlockWidget> block,
      Point xBounds,
      Point yBounds,
      Color backgroundColor,
      Symbols.Marker marker,
      Consumer<Context> painter) {
    this.block = block;
    this.xBounds = xBounds;
    this.yBounds = yBounds;
    this.backgroundColor = backgroundColor;
    this.marker = marker;
    this.painter = painter;
  }

  public static CanvasWidget empty(Consumer<Context> painter) {
    return new CanvasWidget(
        Optional.empty(),
        Point.Zero,
        Point.Zero,
        Color.Reset,
        Symbols.Marker.Braille,
        painter);
  }

  public CanvasWidget withBlock(BlockWidget b) {
    return new CanvasWidget(Optional.of(b), xBounds, yBounds, backgroundColor, marker, painter);
  }

  public CanvasWidget withXBounds(Point p) {
    return new CanvasWidget(block, p, yBounds, backgroundColor, marker, painter);
  }

  public CanvasWidget withYBounds(Point p) {
    return new CanvasWidget(block, xBounds, p, backgroundColor, marker, painter);
  }

  public CanvasWidget withBackgroundColor(Color c) {
    return new CanvasWidget(block, xBounds, yBounds, c, marker, painter);
  }

  public CanvasWidget withMarker(Symbols.Marker m) {
    return new CanvasWidget(block, xBounds, yBounds, backgroundColor, m, painter);
  }

  @Override
  public void render(Rect area, Buffer buf) {
    Rect canvasArea;
    if (block.isPresent()) {
      BlockWidget b = block.get();
      Rect innerArea = b.inner(area);
      b.render(area, buf);
      canvasArea = innerArea;
    } else {
      canvasArea = area;
    }

    buf.setStyle(canvasArea, Style.DEFAULT.withBg(backgroundColor));

    Context ctx =
        Context.create(canvasArea.width(), canvasArea.height(), xBounds, yBounds, marker);
    painter.accept(ctx);
    ctx.finish();

    for (Layer layer : ctx.layers) {
      int n = Math.min(layer.string().length(), layer.colors().length);
      Ranges.range(
          0,
          n,
          i -> {
            char ch = layer.string().charAt(i);
            Color color = layer.colors()[i];
            if (ch != ' ' && ch != '⠀') {
              int x = i % canvasArea.width();
              int y = i / canvasArea.width();
              buf.get(x + canvasArea.left(), y + canvasArea.top())
                  .setChar(ch)
                  .setFg(color);
            }
          });
    }

    double left = xBounds.x();
    double right = xBounds.y();
    double top = yBounds.y();
    double bottom = yBounds.x();
    double width = Math.abs(xBounds.y() - xBounds.x());
    double height = Math.abs(yBounds.y() - yBounds.x());
    double resWidth = (double) (canvasArea.width() - 1);
    double resHeight = (double) (canvasArea.height() - 1);
    Ranges.range(
        0,
        ctx.labels.size(),
        i -> {
          Label l = ctx.labels.get(i);
          if (l.x() >= left && l.x() <= right && l.y() <= top && l.y() >= bottom) {
            int x = (int) ((l.x() - left) * resWidth / width) + canvasArea.left();
            int y = (int) ((top - l.y()) * resHeight / height) + canvasArea.top();
            buf.setSpans(x, y, l.spans(), canvasArea.right() - x);
          }
        });
  }
}
